src/TcpAdapterProxy.cpp

```diff
@@ -1639,99 +1912,142 @@ namespace aws { namespace iot { namespace securedtunneling {
-                                    server->connection_->socket() = std::move(new_socket);
-                                    BOOST_LOG_SEV(log, info) << "Accepted tcp connection on port " << server->connection_->socket().local_endpoint().port() << " from " << server->connection_->socket().remote_endpoint();
-                                    invoke_and_clear_handler(server->after_setup_tcp_socket);
-                                }
-                            });
+                            do_accept_tcp_connection(tac, retry_config, service_id, local_port, true);
                         }
                     }
                 }
             });
     }
 
-    void tcp_adapter_proxy::async_resolve_destination_for_connect(tcp_adapter_context &tac, std::shared_ptr<basic_retry_config> retry_config, string const & service_id, boost::system::error_code const &ec, tcp::resolver::results_type results)
+    void tcp_adapter_proxy::do_accept_tcp_connection(tcp_adapter_context &tac, std::shared_ptr<basic_retry_config> retry_config, string service_id, std::uint16_t local_port, bool is_first_connection)
     {
-        BOOST_LOG_SEV(log, trace) << "Resolve destination to connect for service id: " << service_id;
+        retry_config->operation = std::bind(&tcp_adapter_proxy::do_accept_tcp_connection, this, std::ref(tac), retry_config, service_id, local_port, is_first_connection);
+        tcp_server::pointer server = tac.serviceId_to_tcp_server_map[service_id];
+
+        server->acceptor_.async_accept(tac.io_ctx,
+                [=, &tac](boost::system::error_code const &ec, boost::asio::ip::tcp::socket new_socket)
+                {
+                    if (ec)
+                    {
+                        BOOST_LOG_SEV(log, error) << (boost::format("Could not listen/accept incoming connection on %1%:%2% -- %3%")
+                                                      % tac.bind_address_actual % local_port % ec.message()).str();
+                        basic_retry_execute(log, retry_config,
+                                            [=, &ec]() { throw std::runtime_error((boost::format("Failed to accept new connection on %1% -- %2%") % local_port % ec.message()).str()); });
+                    }
+                    else
+                    {
+                        BOOST_LOG_SEV(log, debug) << "socket port " << new_socket.local_endpoint().port();
+                        string endpoint = boost::lexical_cast<std::string>(new_socket.local_endpoint().port());
+                        BOOST_LOG_SEV(log, debug) << "endpoint mapping:";
+                        for (auto m: tac.adapter_config.serviceId_to_endpoint_map)
+                        {
+                            BOOST_LOG_SEV(log, debug) << m.first << " = " << m.second;
+                        }
+                        tcp_server::pointer server = tac.serviceId_to_tcp_server_map[service_id];
+
+                        uint32_t new_connection_id = ++server->highest_connection_id;
```

Comment:
I see that `highest_connection_id` is monotonically increasing and will overflow to 0 at 4B. Since 0 is invalid connection_id, do we have to implement wraparound to 1 instead of 0? 

I know that `MAX_ACTIVE_CONNECTIONS` limits the number of active connections, but it would be possible in some rare case to start-reset enough connections to hit the maximum. If that happens, then without wraparound to 1, we could have a broken connection mapped to 0 or (worse) duplicate connections mapped to 1.

Reply:
I originally made connection id an unsigned long, but after a bit of deliberation with the team, we agreed to keep it as a 32 bit int. The tunnel lifetime imposes a strict time limit, after which the localproxy would shut down. This would require almost 100000 connections to be created per second to overflow within 12 hours